File: code/ex45.java
class Solution {
    public int jump(int[] nums) {       
        int len = nums.length;
        // Special situation
        if (len == 1)   return 0;
        

        int[] final_pos = new int[len - 1];
        
        // Build an array to record the last possible position each jump can reach
        for (int i = 0; i < len - 1; i++) {
            if (nums[i] + i >= len)      final_pos[i] = len - 1;
            else    final_pos[i] = nums[i] + i;
        }
        
        
        int i = 0;
        int jump = 0;
        int max;
        while (true) {
            jump++;
            if (final_pos[i] == len - 1)    break;  
            
            max = i + 1;
            for (int j = i + 1; j <= final_pos[i]; ++j) 
                if (final_pos[j] >= final_pos[max])
                    max = j;
            
            i = max;
        }
         
        
        return jump;
    }
}

File: code/ex45.cpp
class Solution {
public:
    int jump(vector<int>& nums) {
        int n = nums.size();    if (n <= 1)    return 0;
        int jumps = 0, i = 0;
        while (1) {
            int nxtJump = nums[i];
            int nxt = -1, MAX = INT_MIN;
            for (int j = i + 1; j <= min(i + nxtJump, n - 1); j++) {
                if (j == n - 1)    return jumps + 1;
                if (j + nums[j] > MAX) {
                    MAX = j + nums[j];
                    nxt = j;
                }
            }
            i = nxt;
            jumps++;
        }
        return -1;
    }
};